src/algebra/add_submonoid_closure.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import ring_theory.subsemiring
+
+/-!
+#  Equality between `add_submonoid.closure S` and `subsemiring.closure S`
+
+Let `R` be a semiring and let `S` be a multiplicative submonoid of `R`.
+The additive closure of `S` is a subsemiring of `R`.  If `R` is commutative, then so
+is the additive closure of `S`.
+-/
+variables {R : Type*} {a b : R}
+
+section semiring
+
+variables [semiring R] (S : submonoid R)
+
+/-- The product of an element of the additive closure of a multiplicative submonoid `S`
+and an element of `S` is contained in the additive closure of `S`. -/
+lemma submonoid.mul_mem_add_closure (ha : a ∈ add_submonoid.closure (S : set R)) (hb : b ∈ S) :
+  a * b ∈ add_submonoid.closure (S : set R) :=
+begin
+  revert b,
+  refine add_submonoid.closure_induction ha _ _ _; clear ha a,
+  { exact λ r hr b hb, add_submonoid.mem_closure.mpr (λ y hy, hy (S.mul_mem hr hb)) },
+  { exact λ b hb, by simp only [zero_mul, (add_submonoid.closure (S : set R)).zero_mem] },
+  { simp_rw add_mul,
+    exact λ r s hr hs b hb, (add_submonoid.closure (S : set R)).add_mem (hr hb) (hs hb) }
+end
+
+variable {S}
+
+/-- The product of two elements of the additive closure of a submonoid `S` is an element of the
+additive closure of `S`. -/
+lemma mul_mem
+  (ha : a ∈ add_submonoid.closure (S : set R)) (hb : b ∈ add_submonoid.closure (S : set R)) :
+  a * b ∈ add_submonoid.closure (S : set R) :=
+begin
+  revert a,
+  refine add_submonoid.closure_induction hb _ _ _; clear hb b,
+  { exact λ r hr b hb, S.mul_mem_add_closure hb hr },
+  { exact λ b hb, by simp only [mul_zero, (add_submonoid.closure (S : set R)).zero_mem] },
+  { simp_rw mul_add,
+    exact λ r s hr hs b hb, (add_submonoid.closure (S : set R)).add_mem (hr hb) (hs hb) }
+end
+
+/-- The additive closure of a submonoid is a subsemiring. -/
+def submonoid.to_subsemiring (S : submonoid R) : subsemiring R :=
+{ carrier := add_submonoid.closure (S : set R),
```

Comment:
You should add a pair of `rfl` lemmas saying that

* `to_semiring_coe : (S.to_subsemiring : set R) = add_submonoid.closure (S : set R)`
* `to_semiring_coe_add_submonoid : (S.to_subsemiring : add_submonoid R) = add_submonoid.closure (S : set R)`

Reply:
The first one worked as is.  The second one gives an error (red squiggle under the first `S`):

```lean
lemma to_semiring_coe_add_submonoid :
  (S.to_subsemiring : add_submonoid R) = add_submonoid.closure (S : set R) := rfl

invalid type ascription, term has type
  subsemiring R
but is expected to have type
  add_submonoid R
```

Reply:
If I am debugging correctly, I think that the instance below is what should have and gives an error, right?

```lean
instance pro (T : subsemiring R) : add_submonoid R := by apply_instance

apply_instance tactic fail, target is not a type class
state:
R : Type u_1,
_inst_1 : semiring R,
T : subsemiring R
⊢ add_submonoid R
```

Reply:
Indeed, this works!

```lean
lemma to_semiring_coe_add_submonoid :
  S.to_subsemiring.to_add_submonoid = add_submonoid.closure (S : set R) := rfl
```

Reply:
That last comment is what I meant to write!